pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1070,6 +1080,20 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 			return volumeToMount.GenerateError("MapVolume.MapDevice failed", mapErr)
 		}
 
+		// For CSI driver, devicePath should be checked here by checking symlink from
+		// {volumeMapPath}/{volName}.
+		if isCSI {
+			kvh, ok := og.GetVolumePluginMgr().Host.(volume.KubeletVolumeHost)
+			if !ok {
+				return volumeToMount.GenerateError("MapVolume type assertion error", fmt.Errorf("volume host does not implement KubeletVolumeHost interface"))
+			}
+			hu := kvh.GetHostUtil()
+			devicePath, err = hu.EvalHostSymlinks(filepath.Join(volumeMapPath, volName))
+			if err != nil {
+				return volumeToMount.GenerateError("MapVolume.EvalHostSymlinks failed", err)
+			}
+		}
```

Comment:
Also we need to fix when ASOW gets updated after node is staged etc. Currently both device is marked as mounted and volume is marked as mounted at the end. This needs to be split, how filesystem mounting is split.

Reply:
Fixe by moving  `MarkDeviceAsMounted` before `MapDevice`.